### PR TITLE
Creación 04 Production Configuration para Webpack 4

### PR DESCRIPTION
**Pull request by Senesero**
Se ha partido desde el ejemplo "03 hmr" con la versión 4 de webpack y se ha realizado el ejemplo "04 Production Configuration" donde se han separado los archivos de configuración de webpack para diferenciar entre desarrollo y producción.

También para producción se ha configurado "CompressionPlugin" para obtener unos archivos más comprimidos

### CODE REVIEW

**Comment by Senesero** [comment]
Por cierto, soy Pablo Rodríguez de Lemoncode

File: version 4/03 environments/04 Production Configuration/README.md [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+
+## steps
+
+- `npm install` to install previous sample packages:
+
+```
+npm install
+```
+
+- We are going to start with creating a base webpack configuration file where we are going to add the common features for both environments (`dev` and `prod`). One of the feature which we are split between environments is `sourcemaps`. [Webpack recommends](https://webpack.js.org/guides/production/#source-mapping) that we enable source maps for `production` environment, due to are useful for debugging and to run benchmark tests, but we have to disable it for normal use.
```
Comment (brauliodiez) [inline version 4/03 environments/04 Production Configuration/README.md]: Fixing some grammar / wording:

- We are will start by creating base webpack configuration that will hold all the common feature for both environments (_dev_ and _prod). Then we will place specific configuration for dev and production in their config files (e.g. sourcemaps). Webpack, recommends enabling sources maps for development environment (debuggging purposes), but disable it for normal use.

File: version 4/03 environments/04 Production Configuration/README.md [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+- Let's go to rename `webpack.config.js` to `base.webpack.config.js`:
```
Comment (brauliodiez) [inline version 4/03 environments/04 Production Configuration/README.md]: Let's rename

File: version 4/03 environments/04 Production Configuration/README.md [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    }),
+    new MiniCssExtractPlugin({
+      filename: "[name].css",
+      chunkFilename: "[id].css"
+    }),
+  ],
+};
+```
+
+- Now it's time to install `webpack-merge` package. This allow us to combine `base.webpack.config` with environment config:
```
Comment (brauliodiez) [inline version 4/03 environments/04 Production Configuration/README.md]: Now it's time to install `webpack-merge` package. This allow us to combine `base.webpack.config` with dev or production specific config settings:

File: version 4/03 environments/04 Production Configuration/README.md [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+
+```
+npm install webpack-merge --save-dev
+```
+
+- Let's go to create the `dev` environment configuration:
```
Comment (brauliodiez) [inline version 4/03 environments/04 Production Configuration/README.md]: Let's create

File: version 4/03 environments/04 Production Configuration/README.md [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+module.exports = merge(common, {
+    mode: 'development',
+    devtool: 'inline-source-map',
+    devServer: {
+        contentBase: './dist'
+    }
+});
+```
+
+- Make a configuration to see in console the enviroment in which we are
```
Comment (brauliodiez) [inline version 4/03 environments/04 Production Configuration/README.md]: Just to make a quick check we will add console log, to check in which environment we are running the app (dev or production)

File: version 4/03 environments/04 Production Configuration/README.md [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+- We are going to create a `build dev` command to see how much size has bundles files without `webpack-dev-server` stuff:
+
+### ./package.json
+```diff
+{
+  ...
+  "scripts": {
+    "start": "webpack-dev-server --open --config dev.webpack.config.js",
+-    "build": "rimraf dist && webpack --mode development"
++    "build": "rimraf dist && webpack --config dev.webpack.config.js"
```
Comment (brauliodiez) [inline version 4/03 environments/04 Production Configuration/README.md]: missing build:dev ?

File: version 4/03 environments/04 Production Configuration/README.md [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+  ...
+}
+
+```
+
+- Running `npm run build`:
+
+The `dist` folder is created. If we execute `index.html`, it shows us `We are in: development`.
+
+- Let's configure `production` environment:
```
Comment (brauliodiez) [inline version 4/03 environments/04 Production Configuration/README.md]: Let's configure the

File: version 4/03 environments/04 Production Configuration/README.md [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+const merge = require('webpack-merge');
+const common = require('./base.webpack.config.js');
+
+module.exports = merge(common, {
+    mode: 'production',
+});
+
+```
+
+- Add production command script:
```
Comment (brauliodiez) [inline version 4/03 environments/04 Production Configuration/README.md]: Let's add a production command script

File: version 4/03 environments/04 Production Configuration/README.md [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+-    "build": "rimraf dist && webpack --config dev.webpack.config.js"
++    "build": "rimraf dist && webpack --config dev.webpack.config.js",
++    "build:prod": "rimraf dist && webpack --config prod.webpack.config.js"
+  },
+  ...
+}
+
+```
+
+- Running `npm run build:prod`, we can see in console `We are in: production` and we can see how all bundle sizes decrease:
```
Comment (brauliodiez) [inline version 4/03 environments/04 Production Configuration/README.md]: If we runn `npm run build:prod`, the following output will be dumped into the console log `We are in: production` and we can check that all bundle sizes decrease:

File: version 4/03 environments/04 Production Configuration/README.md [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+  0.css           => 170 KB
+  app.js          =>  12 KB
+  app.css         =>   1 KB
+  index.html      =>   1 KB
+  vendor.css      => 108 KB
+  vendorStyles.js =>   2 KB
+
+```
+
+- As an optional configuration, we can go one over step and generate gzipped bundles. We need install [compression-webpack-plugin](https://github.com/webpack-contrib/compression-webpack-plugin):
```
Comment (brauliodiez) [inline version 4/03 environments/04 Production Configuration/README.md]: Just as an optional extra step, we could generated the gzipped bundles (usually this is just handled by the web servers but is a good idea to generate this files in order to check the size). We need to install [compression-webpack-plugin](https://github.com/webpack-contrib/compression-webpack-plugin):

File: version 4/03 environments/04 Production Configuration/README.md [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+
+```
+npm install compression-webpack-plugin --save-dev
+```
+
+- Updating `prod.webpack.config`:
```
Comment (brauliodiez) [inline version 4/03 environments/04 Production Configuration/README.md]: Let's update

File: version 4/03 environments/04 Production Configuration/README.md (hunk 1 above)
Comment (Senesero) [reply version 4/03 environments/04 Production Configuration/README.md]: done

File: version 4/03 environments/04 Production Configuration/README.md (hunk 2 above)
Comment (Senesero) [reply version 4/03 environments/04 Production Configuration/README.md]: done

File: version 4/03 environments/04 Production Configuration/README.md (hunk 3 above)
Comment (Senesero) [reply version 4/03 environments/04 Production Configuration/README.md]: done

File: version 4/03 environments/04 Production Configuration/README.md (hunk 4 above)
Comment (Senesero) [reply version 4/03 environments/04 Production Configuration/README.md]: done

File: version 4/03 environments/04 Production Configuration/README.md (hunk 5 above)
Comment (Senesero) [reply version 4/03 environments/04 Production Configuration/README.md]: done

File: version 4/03 environments/04 Production Configuration/README.md (hunk 6 above)
Comment (Senesero) [reply version 4/03 environments/04 Production Configuration/README.md]: done

File: version 4/03 environments/04 Production Configuration/README.md (hunk 7 above)
Comment (Senesero) [reply version 4/03 environments/04 Production Configuration/README.md]: done

File: version 4/03 environments/04 Production Configuration/README.md (hunk 8 above)
Comment (Senesero) [reply version 4/03 environments/04 Production Configuration/README.md]: done

File: version 4/03 environments/04 Production Configuration/README.md (hunk 9 above)
Comment (Senesero) [reply version 4/03 environments/04 Production Configuration/README.md]: done

File: version 4/03 environments/04 Production Configuration/README.md (hunk 11 above)
Comment (Senesero) [reply version 4/03 environments/04 Production Configuration/README.md]: done

**Pull request merged**